File: services/ddil-engine/app/network_graph.py
"""
Network graph computation and partition detection.

Builds the adjacency graph for the swarm and identifies connected components.
"""

import numpy as np
from .link_model import link_quality


def compute_network(
    vehicles: list[dict],
    buildings: list[dict] | None = None,
    jammers: list[dict] | None = None,
) -> dict:
    """Compute full network state for the fleet.

    Args:
        vehicles: list of {id, position, comms_range}
        buildings: list of {center, size}
        jammers: list of {center, radius_m, active}

    Returns:
        dict with edges, partitions, partition_count
    """
    edges = []
    vehicle_ids = [v['id'] for v in vehicles]

    for i, v1 in enumerate(vehicles):
        for j, v2 in enumerate(vehicles):
            if i >= j:
                continue

            pos1 = np.array(v1['position'], dtype=float)
            pos2 = np.array(v2['position'], dtype=float)

            # Use minimum comms range (conservative estimate)
            max_range = min(v1.get('comms_range', 800), v2.get('comms_range', 800))

            quality = link_quality(pos1, pos2, max_range, buildings, jammers)
            active = quality > 0.1
            latency = 10.0 / max(quality, 0.01) if active else 0.0

            edges.append({
                'src': v1['id'],
                'dst': v2['id'],
                'quality': round(quality, 4),
                'latency_ms': round(latency, 1),
                'active': active,
            })

    # Find partitions
    active_edges = [(e['src'], e['dst']) for e in edges if e['active']]
    partitions = find_partitions(vehicle_ids, active_edges)

    return {
        'edges': edges,
        'partitions': partitions,
        'partition_count': len(partitions),
    }


def find_partitions(nodes: list[str], edges: list[tuple[str, str]]) -> list[list[str]]:
    """Find connected components using union-find."""
    parent = {n: n for n in nodes}

    def find(x):
        while parent[x] != x:
            parent[x] = parent[parent[x]]
            x = parent[x]
        return x

    def union(a, b):
        ra, rb = find(a), find(b)
        if ra != rb:
            parent[ra] = rb

    for a, b in edges:
        if a in parent and b in parent:
            union(a, b)

    components: dict[str, list[str]] = {}
    for n in nodes:
        root = find(n)
        components.setdefault(root, []).append(n)

    return list(components.values())

File: apps/replay-ui/src/sim/ddil/network-graph.ts
/**
 * Network graph computation and partition detection.
 * Ported from services/ddil-engine/app/network_graph.py
 */

import { linkQuality, type Vec3, type Building, type JammerDef } from "./link-model";

export interface NetworkEdge {
  src: string;
  dst: string;
  quality: number;
  latency_ms: number;
  active: boolean;
}

export interface NetworkResult {
  edges: NetworkEdge[];
  partitions: string[][];
  partition_count: number;
}

export interface VehiclePos {
  id: string;
  position: Vec3;
  comms_range: number;
}

/** Union-Find with path compression for partition detection. */
export function findPartitions(nodes: string[], edges: [string, string][]): string[][] {
  const parent: Record<string, string> = {};
  for (const n of nodes) parent[n] = n;

  function find(x: string): string {
    while (parent[x] !== x) {
      parent[x] = parent[parent[x]]; // path compression
      x = parent[x];
    }
    return x;
  }

  for (const [a, b] of edges) {
    if (a in parent && b in parent) {
      const ra = find(a), rb = find(b);
      if (ra !== rb) parent[ra] = rb;
    }
  }

  const components: Record<string, string[]> = {};
  for (const n of nodes) {
    const root = find(n);
    if (!components[root]) components[root] = [];
    components[root].push(n);
  }

  return Object.values(components);
}

/** Compute full network state for the fleet. */
export function computeNetwork(
  vehicles: VehiclePos[],
  buildings: Building[] = [],
  jammers: JammerDef[] = [],
): NetworkResult {
  const edges: NetworkEdge[] = [];
  const vehicleIds = vehicles.map(v => v.id);

  for (let i = 0; i < vehicles.length; i++) {
    for (let j = i + 1; j < vehicles.length; j++) {
      const v1 = vehicles[i], v2 = vehicles[j];
      const maxRange = Math.min(v1.comms_range, v2.comms_range);
      const quality = linkQuality(v1.position, v2.position, maxRange, buildings, jammers);
      const active = quality > 0.1;
      const latency_ms = active ? Math.round(10 / Math.max(quality, 0.01) * 10) / 10 : 0;

      edges.push({ src: v1.id, dst: v2.id, quality: Math.round(quality * 10000) / 10000, latency_ms, active });
    }
  }

  const activeEdges: [string, string][] = edges.filter(e => e.active).map(e => [e.src, e.dst]);
  const partitions = findPartitions(vehicleIds, activeEdges);

  return { edges, partitions, partition_count: partitions.length };
}
